### ISSUE TITLE
Nest 8 + redis bug

### ISSUE THREAD

**Issue by herenickname** [skunight/nestjs-redis#82]
[Nest] 2697522  - 07/09/2021, 12:43:37 AM   ERROR [ExceptionHandler] Nest can't resolve dependencies of the RedisCoreModule (Symbol(REDIS_MODULE_OPTIONS), ?). Please make sure that the argument ModuleRef at index [1] is available in the RedisCoreModule context.

Potential solutions:
- If ModuleRef is a provider, is it part of the current RedisCoreModule?
- If ModuleRef is exported from a separate @Module, is that module imported within RedisCoreModule?
  @Module({
    imports: [ /* the Module containing ModuleRef */ ]
  })

**Comment by aoxiang78** [comment]
Stopped here, there is no way to continue, do you have a temporary solution?

**Comment by rares-lupascu** [comment]
i personally duplicated this module into my local nestjs project

**Comment by aoxiang78** [comment]
@rares-lupascu Which files have you updated? Can you share it? I haven't solved this problem

**Comment by rares-lupascu** [comment]
@aoxiang78  ok might sound weird but just add this to `redis-core.module.ts`:
1) `import { ConfigModule } from '@nestjs/config';`
2) in the module definition .... change this
```
@Module({
  providers: [RedisService],
  exports: [RedisService],
})
```
to this (line 20):
```
@Module({
    providers: [RedisService],
    exports: [RedisService],
    imports: [ConfigModule],
})
```

**Comment by aoxiang78** [comment]
Solved it your way, thank you for your help @rares-lupascu

**Comment by brandongit2** [comment]
@rares-lupascu That doesn't seem to be working for me. Could you explain how adding `ConfigModule` to the imports solves NestJS looking for `ModuleRef`? That part confuses me.

**Comment by rares-lupascu** [comment]
@brandongit2 what I did was to create a local Redis module copying the contents of this repo into it (the lib folder) and applying the change I did at the top ... hope this helps

**Comment by alikil** [comment]
I cannot use this solution at production build.
Still waiting for this library fix.

**Comment by liaoliaots** [comment]
[@liaoliaots/nestjs-redis v3.0.0-next.2 has been released.](https://www.npmjs.com/package/@liaoliaots/nestjs-redis/v/3.0.0-next.2)

**Comment by PraveenKumarPalai** [comment]
@rares-lupascu Nice workaround for local 😎 !!

But for production, when can we expect a fix on this?

**Comment by bangbang93** [comment]
Keep watching. We are migrating to nestjs 8 and stuck on this.

**Comment by GyanendroKh** [comment]
Instead of using `ModuleRef` to get the `RedisClient`, the `RedisClient` could be directly injected using `@Inject(REDIS_CLIENT)`. And this solves the problem. I opened a pull request #85. 

For now I'm using the library by directly downloading it from my Github [Repo](https://github.com/GyanendroKh/nestjs-redis.git). Hope it will be released soon.
```json
{
  "dependencies": {
    "nestjs-redis": "git+https://github.com/GyanendroKh/nestjs-redis.git#nest8-fix",
  }
}
```

**Comment by cdlaiqi** [comment]
@rares-lupascu solved the problem, thank you

**Comment by nolawnchairs** [comment]
It's been over two years since the last release was published, so I personally consider this project abandoned.

**Comment by SampaioLeal** [comment]
> It's been over two years since the last release was published, so I personally consider this project abandoned.

you mean the 1.3.3 4 months ago?

**Comment by Yurko-Fedoriv** [comment]
Adding here: actually, no code change is required, and lib just needs to be recompiled with updated dependencies.
If you simply copy `lib/ ` into your src codebase and import from it instead of `nestjs-redis` app will work just fine.

That is why [this "weird" solution](https://github.com/skunight/nestjs-redis/issues/82#issuecomment-877677986
) works
And while this [PR](https://github.com/skunight/nestjs-redis/issues/82#issuecomment-886147579) is valid, it is not needed too.

If you are using yarn, you probably can just install the package from github and import from `lib/`
npm will try to run [prepublish](https://github.com/skunight/nestjs-redis/blob/master/package.json#L14) and fail on non-existing yarn.
You can fiddle with postinstall command of your project to pull the dependency from gitlab.

I personally will just copy the `lib/` into my codebase and hope the package will eventually get the update.

**Comment by zackarychapple** [comment]
> git+https://github.com/skunight/nestjs-redis.git

Confirmed that this worked for me as well.  Hope to see this updated as using git+https is not as reliable for protecting against breaking changes.

**Comment by volodyad** [comment]
> > git+https://github.com/skunight/nestjs-redis.git
> 
> Confirmed that this worked for me as well. Hope to see this updated as using git+https is not as reliable for protecting against breaking changes.

You could connect to git commit   "nestjs-redis": "git+https://github.com/skunight/nestjs-redis.git#939f33a7f9b252e872c6edb76a3a860ed053bacf", it will be safer at least

**Comment by pietrovismara** [comment]
@skunight any change of getting this fix in the latest on npm?

**Comment by Distortedlogic** [comment]
bump

**Comment by robertsLando** [comment]
Solution: `npm install @liaoliaots/nestjs-redis`

**Comment by Distortedlogic** [comment]
@robertsLando `src/index.ts(8,30): error TS2307: Cannot find module 'nestjs-redis' or its corresponding type declarations.` uors did not work for me :/

**Comment by robertsLando** [comment]
@Distortedlogic you have to import `@liaoliaots/nestjs-redis` not `nestjs-redis`

**Comment by stefan030** [comment]
> @Distortedlogic you have to import `@liaoliaots/nestjs-redis` not `nestjs-redis`

Still getting similar error after trying out liaoliaots' package. 

Nest can't resolve dependencies of the Symbol(create-nestjs-middleware-module:options) (?, ConfigService). Please make sure that the argument RedisService at index [0] is available in the CoreModule context.

My nestjs version is 8.2.6

**Comment by stefan030** [comment]
As well as: 
The redis client-provider 'Symbol(default)' was not found in the application context.

**Comment by Distortedlogic** [comment]
@stefan030 read the docs... this redis module cannot just be dropped inplace of the other. The config options have a different structure.. also, the module is global only! so u cannot import RedisModule into other modules, instead since its global just inject the service.

**Comment by robertsLando** [comment]
TLDR;:

```
import { RedisModule } from '@liaoliaots/nestjs-redis';

/* ....  /* 
 RedisModule.forRoot({
      closeClient: true,
      config: {
        host: 'localhost',
        port: 6380,
      },
    }),

```

**Comment by JustDoItSascha** [comment]
No release yet?

**Comment by Halinen** [comment]
尝试用yarn更新了半天，最后直接把fix的版本clone了下来解决了- -

**Comment by JuKu** [comment]
@robertsLando Thank you very much! :)

**Comment by ymc9** [comment]
It turns out that for liaoliaots package RedisModule is by default global, you'll run into error if you import it into another module...  Remove the import and the error is gone.

**Comment by Celleb** [comment]
> @rares-lupascu Nice workaround for local 😎 !!
> 
> But for production, when can we expect a fix on this?

You can use  https://www.npmjs.com/package/patch-package

```shell
npx patch-package nestj-redis
```

This will apply your changes every time you install or reinstall the package

**Comment by phillippelevidad** [comment]
> This issue is still exist while installing with npm. Looks like you need to republish your npm package. Currently, I'm overcome this issue by changing the package.json
> 
> ```
> "nestjs-redis": "git+https://github.com/skunight/nestjs-redis.git",
> ```
> 
> ![image](https://user-images.githubusercontent.com/20408623/137195021-a66b30d8-483b-40e9-ad1d-269c95bbfe26.png)

This still seems to be the best solution in 2024!
